EKF/mag_fusion.cpp

```diff
@@ -858,6 +766,155 @@ void Ekf::fuseHeading()
+
+			}
+		} else {
+			_last_static_yaw = predicted_hdg;
+
+		}
+
+		fuseYaw321(measured_hdg, R_YAW, false);
+
+	} else {
+
+		// pitched more than rolled so use 312 rotation order to calculate the observed yaw angle
+		predicted_hdg = atan2f(-_R_to_earth(0, 1), _R_to_earth(1, 1));
+		if (_control_status.flags.mag_hdg) {
+
+			// Calculate the body to earth frame rotation matrix from the euler angles using a 312 rotation sequence
+			// with yaw angle set to to zero
+			Vector3f rotVec312;
+			rotVec312(0) = 0.0f; // first rotation (yaw) set to zero for alter use when rotating the mag field into earth frame
+			rotVec312(1) = asinf(_R_to_earth(2, 1)); // second rotation (roll)
+			rotVec312(2) = atan2f(-_R_to_earth(2, 0), _R_to_earth(2, 2)); // third rotation (pitch)
+			const Dcmf R_to_earth = taitBryan312ToRotMat(rotVec312);
+
+			// rotate the magnetometer measurements into earth frame using a zero yaw angle
+			if (_control_status.flags.mag_3D) {
+				// don't apply bias corrections if we are learning them
+				mag_earth_pred = R_to_earth * _mag_sample_delayed.mag;
+			} else {
+				mag_earth_pred = R_to_earth * (_mag_sample_delayed.mag - _state.mag_B);
+			}
+
+			// the angle of the projection onto the horizontal gives the yaw angle
+			measured_hdg = -atan2f(mag_earth_pred(1), mag_earth_pred(0)) + getMagDeclination();
+
+		} else if (_control_status.flags.ev_yaw) {
+			// calculate the yaw angle for a 312 sequence
+			// Values from yaw_input_312.c file produced by https://github.com/PX4/ecl/blob/master/matlab/scripts/Inertial%20Nav%20EKF/quat2yaw312.m
+			float Tbn_0_1_neg = 2.0f*(_ev_sample_delayed.quat(0)*_ev_sample_delayed.quat(3)-_ev_sample_delayed.quat(1)*_ev_sample_delayed.quat(2));
+			float Tbn_1_1 = sq(_ev_sample_delayed.quat(0))-sq(_ev_sample_delayed.quat(1))+sq(_ev_sample_delayed.quat(2))-sq(_ev_sample_delayed.quat(3));
+			measured_hdg = atan2f(Tbn_0_1_neg,Tbn_1_1);
+
+		}
+
+		// handle special case where yaw measurement is unavailable
+		if (_yaw_use_inhibit) {
+			// The yaw measurement cannot be trusted but we need to fuse something to prevent a badly
+			// conditioned covariance matrix developing over time.
+			if (!_control_status.flags.vehicle_at_rest) {
+				// Vehicle is not at rest so fuse a zero innovation if necessary to prevent
+				// unconstrained quaterniion variance growth and record the predicted heading
+				// to use as an observation when movement ceases.
+				float sumQuatVar = P(0,0) + P(1,1) + P(2,2) + P(3,3);
+				if (sumQuatVar > 0.01f) {
+					fuseYaw312(0.0f, 0.25f, true);
+				}
+				_last_static_yaw = predicted_hdg;
+				return;
+
+			} else {
+				// Vehicle is at rest so use the last moving prediction as an observation
+				// to prevent the heading from drifting and to enable yaw gyro bias learning
+				// before takeoff.
+				measured_hdg = _last_static_yaw;
+
+			}
+		} else {
+			_last_static_yaw = predicted_hdg;
+
+		}
+
+		fuseYaw312(measured_hdg, R_YAW, false);
+
```

Comment:
```suggestion

		// handle special case where yaw measurement is unavailable
		bool fuse_zero_innov = false;
		if (_yaw_use_inhibit) {
			// The yaw measurement cannot be trusted but we need to fuse something to prevent a badly
			// conditioned covariance matrix developing over time.
			if (!_control_status.flags.vehicle_at_rest) {
				// Vehicle is not at rest so fuse a zero innovation if necessary to prevent
				// unconstrained quaternion variance growth and record the predicted heading
				// to use as an observation when movement ceases.
				const float sumQuatVar = P(0,0) + P(1,1) + P(2,2) + P(3,3);
				if (sumQuatVar > 0.01f) {
					measured_hdg = 0.f
					fuse_zero_innov = true;
					R_YAW = 0.25f;
				}
				_last_static_yaw = predicted_hdg;

			} else {
				// Vehicle is at rest so use the last moving prediction as an observation
				// to prevent the heading from drifting and to enable yaw gyro bias learning
				// before takeoff.
				measured_hdg = _last_static_yaw;

			}
		} else {
			_last_static_yaw = predicted_hdg;

		}

		fuseYaw312(measured_hdg, R_YAW, fuse_zero_innov);

```